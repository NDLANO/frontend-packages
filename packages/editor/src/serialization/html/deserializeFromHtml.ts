/**
 * Copyright (c) 2025-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { jsx as slatejsx } from "slate-hyperscript";
import { Element, Node, Text, type Descendant } from "slate";
import type { ElementType } from "../../types";
import { SECTION_ELEMENT_TYPE } from "../../plugins/section/sectionTypes";
import { PARAGRAPH_ELEMENT_TYPE, type ParagraphElement } from "../../plugins/paragraph/paragraphTypes";
import type { SlateSerializer } from "../../core";
import { isElementOfType } from "../../utils/isElementType";
import { NOOP_ELEMENT_TYPE } from "../../plugins/noop/noopTypes";

// TODO: This entire file should be refactored and reconsidered. Our current deserialization is too complex.

const DEFAULT_NOOP: Descendant[] = [{ type: NOOP_ELEMENT_TYPE, children: [{ text: "" }] }];

const EMPTY_VALUE: Descendant[] = [
  {
    type: SECTION_ELEMENT_TYPE,
    children: [
      {
        type: PARAGRAPH_ELEMENT_TYPE,
        children: [
          {
            text: "",
          },
        ],
      },
    ],
  },
];

interface DeserializeOptions {
  noop?: boolean;

  blocks: ElementType[];
  inlines: ElementType[];
}

export const deserializeFromHtml = (
  html: string,
  rules: SlateSerializer[],
  options: DeserializeOptions,
): Descendant[] => {
  if (!html) {
    return options.noop ? DEFAULT_NOOP : EMPTY_VALUE;
  }
  const deserialize = (el: HTMLElement | ChildNode): Descendant | Descendant[] => {
    if (el.nodeType === 3) {
      return { text: el.textContent || "" };
    } else if (el.nodeType !== 1) {
      return { text: "" };
    }

    let children = Array.from(el.childNodes).flatMap(deserialize);
    if (!children.length) {
      children = [{ text: "" }];
    }

    for (const rule of rules) {
      if (rule.deserialize) {
        // Already checked that nodeType === 1 -> el must be of type HTMLElement.
        const ret = rule.deserialize(el as HTMLElement, children, rule.options);
        if (ret === undefined) {
          continue;
        } else {
          return ret;
        }
      }
    }

    return children;
  };

  const document = new DOMParser().parseFromString(
    options.noop ? `<div data-noop="true">${html}</div>` : html,
    "text/html",
  );
  const nodes = Array.from(document.body.children).map(deserialize);

  const normalizedNodes = nodes
    .map((n) => {
      const node = Node.isNodeList(n) ? n[0] : n;
      return node ? wrapMixedChildren(node, options.blocks, options.inlines) : undefined;
    })
    .filter((n) => !!n);
  return normalizedNodes;
};

const addEmptyTextNodes = (node: Element) => {
  const children = node.children;
  let lastWasText = false;

  // Iterating in reverse ensures that we add empty text nodes only when necessary
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    const currentIsText = Text.isText(child);

    if (!currentIsText && !lastWasText) {
      children.splice(i, 0, { text: "" });
      i++; // Skip next iteration since we inserted a new child
    }
    lastWasText = currentIsText;
  }

  // Ensure the last child is a text node
  if (!Text.isText(children[children.length - 1])) {
    children.push({ text: "" });
  }
};

const addEmptyParagraphs = (node: Element, blocks: ElementType[]) => {
  const children = node.children;
  let lastWasBlock = false;

  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    const currentIsBlock = isElementOfType(child, blocks);

    if (currentIsBlock && (i === 0 || lastWasBlock)) {
      children.splice(i, 0, { type: "paragraph", children: [{ text: "" }] });
      i++; // Skip next iteration since we inserted a new paragraph
    }

    lastWasBlock = currentIsBlock;
  }

  // Ensure the last child is a paragraph if needed
  if (isElementOfType(children[children.length - 1], blocks)) {
    children.push({ type: "paragraph", children: [{ text: "" }] });
  }
};

const wrapMixedChildren = (node: Descendant, blocks: ElementType[], inlines: ElementType[]): Descendant => {
  if (!Element.isElement(node)) return node;
  const children = node.children;

  const blockChildren = children.filter((child) => Element.isElement(child) && !inlines.includes(child.type));
  const mixed = !!blockChildren.length && blockChildren.length !== children.length;
  if (!mixed) {
    node.children = children.map((child) => wrapMixedChildren(child, blocks, inlines));
    if (!blockChildren.length && !!children.length) {
      addEmptyTextNodes(node);
    } else {
      addEmptyParagraphs(node, blocks);
    }
    return node;
  }

  // Handle mixed inline-block content
  const cleanNodes: Descendant[] = [];
  let openWrapperBlock: ParagraphElement | null = null;
  for (const child of children) {
    if (Text.isText(child) || (Element.isElement(child) && inlines.includes(child.type))) {
      // TODO: Consider trimming
      if (Node.string(child) === "" || Node.string(child) === " ") {
        continue;
      }
      if (!openWrapperBlock) {
        openWrapperBlock = slatejsx("element", { type: "paragraph" }, []) as ParagraphElement;
        cleanNodes.push(openWrapperBlock);
      }
      openWrapperBlock.children.push(child);
    } else {
      openWrapperBlock = null;
      if (child.type === "paragraph" && child.children.length === 0) continue;
      cleanNodes.push(child);
    }
  }

  addEmptyParagraphs(node, blocks);

  // Process the cleaned-up nodes recursively
  node.children = cleanNodes.map((child) => wrapMixedChildren(child, blocks, inlines));
  return node;
};
