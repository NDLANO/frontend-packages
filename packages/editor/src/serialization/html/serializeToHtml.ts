/**
 * Copyright (c) 2025-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { Element, type Descendant } from "slate";
import type { SlateSerializer } from "../../types";
import { commonSerializers, extendedSerializers } from "./htmlSerializers";

const serialize = (node: Descendant, rules: SlateSerializer[]): string | undefined => {
  const children = Element.isElement(node)
    ? node.children
        .map((n) => serialize(n, rules))
        .filter((n) => !!n)
        .join("")
    : "";

  for (const rule of rules) {
    if (!rule.serialize) {
      continue;
    }

    const ret = rule.serialize(node, children);
    if (ret === undefined) {
      continue;
    } else if (ret === null) {
      return undefined;
    } else return ret;
  }

  return children;
};

const DELETE_REGEXP = /<deleteme><\/deleteme>/g;

export const serializeToHtml = (value: Descendant[], rules: SlateSerializer[]) => {
  const elements = value
    .map((descendant) => serialize(descendant, rules))
    .filter((n) => !!n)
    .join("");

  return elements.replace(DELETE_REGEXP, "");
};

export const inlineContentToHTML = (value: Descendant[]) => {
  return serializeToHtml(value, commonSerializers);
};

export const blockContentToHTML = (value: Descendant[]) => {
  return serializeToHtml(value, extendedSerializers);
};
