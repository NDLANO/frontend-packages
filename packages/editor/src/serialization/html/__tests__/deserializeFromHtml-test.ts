/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { inlineContentToEditorValue, blockContentToEditorValue } from "../deserializeFromHtml";
import { blockContentToHTML, inlineContentToHTML } from "../serializeToHtml";

const contentHTML = `<section><h2>Lorem ipsum</h2></section>`;

const contentHTMLWithSections = `<section><h2>Section 1</h2></section><section><h2>Section 2</h2></section><section><h2>Section 3</h2></section>`;
const mustBeWrappedHtml = `<section><h2>Section 1</h2><aside>Some text that slate wants to delete <div><em>blabla</em></div></aside></section><section><h2>Section 2</h2></section><section><h2>Section 3</h2></section>`;

test("articleContentConverter convert topic article content to and from editorValue", () => {
  // Todo fix test to handle empty text nodes
  const editorValue = inlineContentToEditorValue(contentHTML);
  const html = inlineContentToHTML(editorValue);
  expect(html).toMatchSnapshot();
});

test("articleContentConverter convert learningresource content to and from editorValue", () => {
  const editorValue = blockContentToEditorValue(contentHTML);
  const html = blockContentToHTML(editorValue);
  expect(html).toMatchSnapshot();
});

test("articleContentConverter convert learningresource content", () => {
  const editorValue = blockContentToEditorValue(contentHTML);
  expect(editorValue[0]).toMatchSnapshot();
});

test("articleContentConverter convert learningresource content with multiple sections to and from editorValue", () => {
  const editorValue = blockContentToEditorValue(contentHTMLWithSections);
  const html = blockContentToHTML(editorValue);
  expect(html).toMatchSnapshot();
});

test("articleContentConverter convert article that is a mix of inline and block object by wrapping the inline", () => {
  const editorValue = blockContentToEditorValue(mustBeWrappedHtml);

  const html = blockContentToHTML(editorValue);
  expect(html).toMatchSnapshot();
});
