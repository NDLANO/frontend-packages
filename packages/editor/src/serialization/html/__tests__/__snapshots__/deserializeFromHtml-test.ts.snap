// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`articleContentConverter convert article that is a mix of inline and block object by wrapping the inline 1`] = `"<section><h2>Section 1</h2>Some text that slate wants to delete <em>blabla</em></section><section><h2>Section 2</h2></section><section><h2>Section 3</h2></section>"`;

exports[`articleContentConverter convert learningresource content 1`] = `
{
  "children": [
    {
      "children": [
        {
          "text": "Lorem ipsum",
        },
      ],
      "level": 2,
      "type": "heading",
    },
  ],
  "type": "section",
}
`;

exports[`articleContentConverter convert learningresource content to and from editorValue 1`] = `"<section><h2>Lorem ipsum</h2></section>"`;

exports[`articleContentConverter convert learningresource content with multiple sections to and from editorValue 1`] = `"<section><h2>Section 1</h2></section><section><h2>Section 2</h2></section><section><h2>Section 3</h2></section>"`;

exports[`articleContentConverter convert topic article content to and from editorValue 1`] = `"<section><h2>Lorem ipsum</h2></section>"`;
